src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/SimpleStrandSwitchVariantDetector.java

```diff
@@ -0,0 +1,301 @@
+        Utils.validateArg(clipLength < input.endInAssembledContig - input.startInAssembledContig + 1,
+                            "input alignment to be clipped away: " + input.toPackedString() + "\twith clip length: " + clipLength);
+
+        final Tuple2<SimpleInterval, Cigar> result = computeNewRefSpanAndCigar(input, clipLength, clipFrom3PrimeEnd);
+        return new AlignmentInterval(result._1, input.startInAssembledContig, input.endInAssembledContig - clipLength,
+                result._2, input.forwardStrand, input.mapQual,
+                StructuralVariationDiscoveryArgumentCollection.DiscoverVariantsFromContigsAlignmentsSparkArgumentCollection.MISSING_NM,
+                input.alnScore, input.isFromSplitGapAlignment, input.hasUndergoneOverlapRemoval);
+    }
+
+    @VisibleForTesting
+    static Tuple2<SimpleInterval, Cigar> computeNewRefSpanAndCigar(final AlignmentInterval input, final int clipLength,
+                                                                   final boolean clipFrom3PrimeEnd) {
+        Utils.validateArg(input.cigarAlong5to3DirectionOfContig.getCigarElements().stream().map(CigarElement::getOperator)
+                        .noneMatch(op -> op.equals(CigarOperator.N) || op.isPadding()),
+                "Input alignment contains padding or skip operations, which is currently unsupported: " + input.toPackedString());
+
+        final Tuple3<List<CigarElement>, List<CigarElement>, List<CigarElement>> threeSections = extractCigarElements(input);
+
+        final List<CigarElement> cigarElements = threeSections._2();
+        int idx;
+        final int step, stop;
+        if (clipFrom3PrimeEnd) {
+            idx = cigarElements.size() - 1; step = -1; stop = -1;
+        } else {
+            idx = 0; step = 1; stop = cigarElements.size();
+        }
+
+        Cigar newCigar = new Cigar();
+        int readBasesConsumed = 0, refBasesConsumed = 0;
+        while (stop != idx) {
+            final CigarElement ce = cigarElements.get(idx);
+            if ( ce.getOperator().equals(CigarOperator.HARD_CLIP))
+                continue;
+            if ( ce.getOperator().consumesReadBases() ) {
+                if (readBasesConsumed + ce.getLength() < clipLength)
+                    readBasesConsumed += ce.getLength();
+                else { // enough read bases would be clipped
+
+                    if ( !ce.getOperator().isAlignment() && !ce.getOperator().equals(CigarOperator.I))
+                        throw new GATKException.ShouldNeverReachHereException("Logic error, should not reach here");
+
+                    // dead with cigar first
+                    final List<CigarElement> resultCEs = threeSections._1();
+                    final int a = readBasesConsumed + ce.getLength() - clipLength;
+                    final CigarOperator op = ce.getOperator().isAlignment() ? CigarOperator.M : CigarOperator.S;
+                    if (clipFrom3PrimeEnd) {
```

Comment:
This code might be a little more concise if you put these things in the sublist in the same order all the time, and then reverse the sublist if `clipFrom3PrimeEnd` with `Collections.reverse()`, and then append the sublist to `resultCEs`.

Reply:
done